File: day03/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
)

const (
	puzzleInput = "input.txt"
)

var digitsRegex = regexp.MustCompile(`\d+`)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	var grid [][]byte

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		grid = append(grid, []byte(scanner.Text()))
	}

	sum := 0

	gears := map[int][]int{}

	lim := Vec2{
		x: len(grid[0]),
		y: len(grid),
	}

	buf := make([]Vec2, lim.x*2+2)
	for y, i := range grid {
		matches := digitsRegex.FindAllIndex(i, -1)
		for _, j := range matches {
			left := Vec2{
				x: j[0],
				y: y,
			}
			right := Vec2{
				x: j[1] - 1,
				y: y,
			}
			num, err := strconv.Atoi(string(i[j[0]:j[1]]))
			if err != nil {
				log.Fatalln(err)
			}
			isSym := false
			n := getNeighbors(left, right, lim, buf)
			for _, k := range buf[:n] {
				if sym := grid[k.y][k.x]; isSymbol(sym) {
					isSym = true
					if sym == '*' {
						id := k.y*lim.x + k.x
						gears[id] = append(gears[id], num)
					}
				}
			}
			if isSym {
				sum += num
			}
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	fmt.Println("Part 1:", sum)

	sum2 := 0
	for _, v := range gears {
		if len(v) == 2 {
			sum2 += v[0] * v[1]
		}
	}

	fmt.Println("Part 2:", sum2)
}

type (
	Vec2 struct {
		x, y int
	}
)

func getNeighbors(p1, p2, lim Vec2, buf []Vec2) int {
	if p1.x > p2.x {
		p1, p2 = p2, p1
	}
	if p1.y > p2.y {
		p1.y, p2.y = p2.y, p1.y
	}
	idx := 0
	if y := p1.y - 1; y >= 0 {
		for i := p1.x; i <= p2.x; i++ {
			buf[idx] = Vec2{
				x: i,
				y: y,
			}
			idx++
		}
	}
	if y := p2.y + 1; y < lim.y {
		for i := p1.x; i <= p2.x; i++ {
			buf[idx] = Vec2{
				x: i,
				y: y,
			}
			idx++
		}
	}
	if x := p1.x - 1; x >= 0 {
		for i := p1.y; i <= p2.y; i++ {
			buf[idx] = Vec2{
				x: x,
				y: i,
			}
			idx++
		}
		if y := p1.y - 1; y >= 0 {
			buf[idx] = Vec2{
				x: x,
				y: y,
			}
			idx++
		}
		if y := p2.y + 1; y < lim.y {
			buf[idx] = Vec2{
				x: x,
				y: y,
			}
			idx++
		}
	}
	if x := p2.x + 1; x < lim.x {
		for i := p1.y; i <= p2.y; i++ {
			buf[idx] = Vec2{
				x: x,
				y: i,
			}
			idx++
		}
		if y := p1.y - 1; y >= 0 {
			buf[idx] = Vec2{
				x: x,
				y: y,
			}
			idx++
		}
		if y := p2.y + 1; y < lim.y {
			buf[idx] = Vec2{
				x: x,
				y: y,
			}
			idx++
		}
	}
	return idx
}

func isSymbol(b byte) bool {
	return b != '.' && (b < '0' || b > '9')
}

File: day03/src/main.rs
use once_cell::sync::Lazy;
use regex::Regex;
use std::collections::HashMap;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

static DIGIT_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"\d+").unwrap());

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let grid = {
        let mut grid = Vec::new();
        for line in reader.lines() {
            grid.push(line?);
        }
        grid
    };

    let mut sum = 0;

    let mut gears = HashMap::new();

    let lim = (grid[0].len(), grid.len());

    let mut buf = vec![(0, 0); lim.1 * 2 + 2];

    for (y, i) in grid.iter().enumerate() {
        for j in DIGIT_REGEX.find_iter(i) {
            let left = (j.start(), y);
            let right = (j.end() - 1, y);
            let num = j.as_str().parse::<usize>()?;
            let mut has_sym = false;
            let n = get_neighbors(left, right, lim, &mut buf);
            for &(x, y) in buf.iter().take(n) {
                let sym = grid[y].as_bytes()[x];
                if is_symbol(sym) {
                    has_sym = true;
                    if sym == b'*' {
                        let id = y * lim.0 + x;
                        let entry = gears.entry(id).or_insert(Vec::new());
                        entry.push(num);
                    }
                }
            }
            if has_sym {
                sum += num;
            }
        }
    }
    println!("Part 1: {}", sum);

    let sum2 = gears
        .values()
        .flat_map(|v| match &v[..] {
            [a, b] => Some(a * b),
            _ => None,
        })
        .sum::<usize>();
    println!("Part 2: {}", sum2);
    Ok(())
}

fn get_neighbors(
    p1: (usize, usize),
    p2: (usize, usize),
    lim: (usize, usize),
    buf: &mut [(usize, usize)],
) -> usize {
    let mut idx = 0;
    if p1.1 > 0 {
        let y = p1.1 - 1;
        for i in p1.0..=p2.0 {
            buf[idx] = (i, y);
            idx += 1;
        }
    }
    {
        let y = p2.1 + 1;
        if y < lim.1 {
            for i in p1.0..=p2.0 {
                buf[idx] = (i, y);
                idx += 1;
            }
        }
    }
    if p1.0 > 0 {
        let x = p1.0 - 1;
        for i in p1.1..=p2.1 {
            buf[idx] = (x, i);
            idx += 1;
        }
        if p1.1 > 0 {
            buf[idx] = (x, p1.1 - 1);
            idx += 1;
        }
        let y = p2.1 + 1;
        if y < lim.1 {
            buf[idx] = (x, y);
            idx += 1;
        }
    }
    {
        let x = p2.0 + 1;
        if x < lim.0 {
            for i in p1.1..=p2.1 {
                buf[idx] = (x, i);
                idx += 1;
            }
            if p1.1 > 0 {
                buf[idx] = (x, p1.1 - 1);
                idx += 1;
            }
            let y = p2.1 + 1;
            if y < lim.1 {
                buf[idx] = (x, y);
                idx += 1;
            }
        }
    }
    idx
}

fn is_symbol(b: u8) -> bool {
    return b != b'.' && (b < b'0' || b > b'9');
}
